packages/instant/src/components/buy_button.tsx

```diff
@@ -35,21 +36,38 @@ export class BuyButton extends React.Component<BuyButtonProps> {
     }
     private readonly _handleClick = async () => {
         // The button is disabled when there is no buy quote anyway.
-        if (_.isUndefined(this.props.buyQuote) || _.isUndefined(this.props.assetBuyer)) {
+        const { buyQuote, assetBuyer } = this.props;
+        if (_.isUndefined(buyQuote) || _.isUndefined(assetBuyer)) {
             return;
         }
-        this.props.onClick(this.props.buyQuote);
-        let txnHash;
+
+        let txHash: string | undefined;
+        this.props.onAwaitingSignature(buyQuote);
         try {
-            txnHash = await this.props.assetBuyer.executeBuyQuoteAsync(this.props.buyQuote);
-            const txnReceipt = await web3Wrapper.awaitTransactionSuccessAsync(txnHash);
-            this.props.onBuySuccess(this.props.buyQuote, txnReceipt.transactionHash);
+            txHash = await assetBuyer.executeBuyQuoteAsync(buyQuote);
         } catch (e) {
             if (e instanceof Error && e.message === AssetBuyerError.SignatureRequestDenied) {
-                this.props.onBuyPrevented(this.props.buyQuote, e);
+                this.props.onSignatureDenied(buyQuote, e);
+            } else {
+                throw e;
+            }
+        }
+
+        // Have to let TS know that txHash is definitely defined now
+        if (!txHash) {
```

Comment:
hm i think if we leave out the `string | undefined` on line 44 the type inference will work w/o this

Reply:
this doesn't work unfortunately, gives the error "Variable 'txHash' is used before being assigned"